standalone/examples_test.go

```diff
@@ -0,0 +1,200 @@
+package standalone
+
+import (
+	"encoding/json"
+	"github.com/fullstorydev/grpcui/testing/testdata"
+	"github.com/google/go-cmp/cmp"
+	"testing"
+	"time"
+)
+
+func TestRequest_MarshalUnmarshal(t *testing.T) {
+	tests := []struct {
+		name    string
+		request Request
+	}{
+		{
+			name: "Full request, complex data",
+			request: Request{
+				Timeout: 1 * time.Second,
+				Metadata: []MetadataPair{
+					{
+						"key",
+						"value",
+					},
+				},
+				Data: map[string]interface{}{
+					"field": "fieldValue",
+				},
+			},
+		},
+		{
+			name: "Full request, string data",
+			request: Request{
+				Timeout: 2 * time.Millisecond,
+				Metadata: []MetadataPair{
+					{
+						"key",
+						"value",
+					},
+				},
+				Data: "a string",
+			},
+		},
+		{
+			name: "No timeout",
+			request: Request{
+				Metadata: []MetadataPair{
+					{
+						"key",
+						"value",
+					},
+				},
+				Data: map[string]interface{}{
+					"field": "fieldValue",
+				},
+			},
+		},
+		{
+			name: "no data",
+			request: Request{
+				Timeout: 1 * time.Millisecond,
+				Metadata: []MetadataPair{
+					{
+						"key",
+						"value",
+					},
+				},
+			},
+		},
+		{
+			name: "only timeout",
+			request: Request{
+				Timeout: 24 * time.Hour,
+			},
+		},
+		{
+			name: "only data",
+			request: Request{
+				Data: "some data",
+			},
+		},
+		{
+			name: "only metadata",
+			request: Request{
+				Metadata: []MetadataPair{
+					{
+						"a",
+						"b",
+					},
+				},
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			marshal, err := json.Marshal(test.request)
+			if err != nil {
+				t.Fatalf("marshal failed: %v", err)
+			}
+			t.Logf("marshaled: %q", string(marshal))
+
+			var unmarshaled Request
+			err = json.Unmarshal(marshal, &unmarshaled)
+			if err != nil {
+				t.Fatalf("unmarshal failed: %v", err)
+			}
+
+			if diff := cmp.Diff(test.request, unmarshaled); diff != "" {
+				t.Fatalf("OSRM gateway request mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestRequest_MarshalJSON_ProtoData(t *testing.T) {
+	tests := []struct {
+		name    string
+		request Request
+	}{
+		{
+			name: "proto3 data",
+			request: Request{
+				Data: testmodels.TestMessage3{
+					AInt32:  107,
+					AString: "string",
+				},
+			},
+		},
+		{
+			name: "proto2 data",
+			request: Request{
+				Data: testmodels.TestMessage2{
```

Comment:
I think we don't even need any custom test protos here -- you could instead test using well-known protos, like `FieldDescriptorProto` and `Struct`.

Reply:
Can you please suggest an example?

Reply:
I did: `FileDescriptorProto` (in `github.com/golang/protobuf/protoc-gen-go/descriptor`) and `Struct` (in `github.com/golang/protobuf/ptypes/struct`):
```go
// maybe something like so
Data: &descriptor.FileDescriptorProto{
    Name: proto.String("foo.proto"),
    MessageType: []*descriptor.DescriptorProto{
        {
            Name: proto.String("Foo"),
        },
    },
}
```

Also, looking more closely at this, it's not clear why this works since `testmodels.TestMessage2` is not a `proto.Message`, the _pointer_ to it (`*testmodels.TestMessage2`) is. I think this ends up using `encoding/json` -- which works if you are just trying to round-trip the value to JSON and back, but does not work to send a message to the browser (`encoding/json` may not encode it in a way that the browser client accepts).

Anyhow, use of a well-known type just keeps this repo a little smaller, leaning on existing things instead of creating new ones... Not a blocking comment, as long as it leads to no extra exported API.

Reply:
Works for me I'll delete the custom types and use `descriptor.DescriptorProto`

Reply:
Done